src/Illuminate/Database/Query/Builder.php

```diff
@@ -2582,11 +2583,15 @@ public function update(array $values)
      */
     public function updateOrInsert(array $attributes, array $values = [])
     {
-        if (! $this->where($attributes)->exists()) {
-            return $this->insert(array_merge($attributes, $values));
-        }
+        DB::transaction(function () use ($attributes, $values) {
```

Comment:
This must be wrong?

There can be multiple database connection and `DB::transaction` just picks the default/first one but the query in question can run on a different one.

Reply:
What do you mean? I suggested that this code should be all in one DB transaction, since `! $this->where($attributes)->exists()` can be no longer true when `$this->insert(array_merge($attributes, $values))` is ran.

Reply:
He means that the query in question can be made on a specific connection while `DB::transaction()` will start the transaction on the default connection. This means that instead of doing `DB::transaction()` you should do `$this->connection->transaction()` since the actual connection gets passed to the `Builder` constructor.

Reply:
I see. 

I mentioned other functions with similar behavior, is my observation of a possible race condition valid? I can open an issue instead.